pkg/api/helpers/upload.go

```diff
@@ -86,16 +99,193 @@ func ClientUpload(ctx context.Context, client apigen.ClientWithResponsesInterfac
+	}
+	if resp.StatusCode() != http.StatusNoContent {
+		return fmt.Errorf("abort presign multipart upload: %w", ResponseAsError(resp))
+	}
+	return nil
+}
+
+func (u *PreSignUploader) init(ctx context.Context) (*apigen.PresignMultipartUpload, error) {
+	// create presign multipart upload
+	resp, err := u.Client.CreatePresignMultipartUploadWithResponse(ctx, u.RepoID, u.BranchID, &apigen.CreatePresignMultipartUploadParams{
+		Path:  u.ObjectPath,
+		Parts: swag.Int(u.NumParts),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("create presign multipart upload: %w", err)
+	}
+	if resp.JSON201 == nil {
+		return nil, fmt.Errorf("create presign multipart upload: %w", ResponseAsError(resp))
+	}
+
+	// verify we got the expected number of presigned urls
+	mpu := resp.JSON201
+	var presignedUrls []string
+	if mpu.PresignedUrls != nil {
+		presignedUrls = *mpu.PresignedUrls
+	}
+	if len(presignedUrls) != u.NumParts {
+		return nil, fmt.Errorf("create presign multipart upload: %w, expected %d presigned urls, got %d", ErrRequestFailed, u.NumParts, len(presignedUrls))
+	}
+	return mpu, nil
+}
+
+func (u *PreSignUploader) uploadPart(ctx context.Context, partReader *io.SectionReader, partURL string, contentType string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, partURL, partReader)
+	if err != nil {
+		return "", err
+	}
+	req.ContentLength = partReader.Size()
+	if contentType != "" {
+		req.Header.Set("Content-Type", contentType)
+	}
+
+	resp, err := u.HTTPClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer func() { _ = resp.Body.Close() }()
+	if !httputil.IsSuccessStatusCode(resp) {
+		return "", fmt.Errorf("upload %w %s: %s", ErrRequestFailed, partURL, resp.Status)
+	}
+
+	etag := extractEtagFromResponseHeader(resp.Header)
+	if etag == "" {
+		return "", fmt.Errorf("upload %w %s: etag is missing", ErrRequestFailed, partURL)
+	}
+	return etag, nil
+}
+
+func ClientUploadPreSign(ctx context.Context, client apigen.ClientWithResponsesInterface, repoID, branchID, objPath string, metadata map[string]string, contentType string, contents io.ReadSeeker, presignMultipartSupport bool) (*apigen.ObjectStats, error) {
 	// calculate size using seek
 	contentLength, err := contents.Seek(0, io.SeekEnd)
 	if err != nil {
 		return nil, err
 	}
 
+	var uploader *PreSignUploader
+	if presignMultipartSupport && canUseMultipartUpload(contents, contentLength) {
+		uploader = NewPreSignUploader(client, repoID, branchID, objPath, contents.(io.ReaderAt), contentLength, contentType, metadata)
+	}
+
 	// upload loop, retry on conflict
+	var stats *apigen.ObjectStats
 	for {
-		stats, err := clientUploadPreSignHelper(ctx, client, repoID, branchID, objPath, metadata, contentType, contents, contentLength)
+		if uploader != nil {
+			stats, err = uploader.Upload(ctx)
+		} else {
+			stats, err = clientUploadPreSignHelper(ctx, client, repoID, branchID, objPath, metadata, contentType, contents, contentLength)
+		}
```

Comment:
This is confusing. `uploader` is `PreSignUploader`. If it's nil, use `clientUploadPreSignHelper` instead. Can you rename vars or use comment to help clarify?

Reply:
Sure, I'll update the name or integrate them together.

Reply:
Integrated presign upload of single object into uploader functionality.